{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMWkd+UNObXs9ZKJXr9Zb5S"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["!pip install biopython"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fDqDfhRB4Kp_","executionInfo":{"status":"ok","timestamp":1678975449186,"user_tz":-60,"elapsed":10580,"user":{"displayName":"lateefat arogundade","userId":"11826085879239467444"}},"outputId":"c01a8ef7-ac17-4f0c-dc99-db4b8da3013c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting biopython\n","  Downloading biopython-1.81-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m19.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.9/dist-packages (from biopython) (1.22.4)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.81\n"]}]},{"cell_type":"code","source":["from Bio import SeqIO"],"metadata":{"id":"yproN48x4TXO"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def extract_promoter_sequences(gff_file, fasta_file, upstream_length, output_file):\n","    # Read in the genome assembly file and store sequences in a dictionary\n","    genome_dict = {}\n","    with open(fasta_file) as f:\n","        sequence = \"\"\n","        header = \"\"\n","        for line in f:\n","            if line.startswith(\">\"):\n","                if header != \"\":\n","                    genome_dict[header] = sequence\n","                header = line.strip()[1:]\n","                sequence = \"\"\n","            else:\n","                sequence += line.strip()\n","        genome_dict[header] = sequence\n","    \n","    # Create a dictionary to store the start and end positions of each gene\n","    gene_dict = {}\n","    with open(gff_file) as f:\n","        for line in f:\n","            if not line.startswith(\"#\"):\n","                fields = line.strip().split(\"\\t\")\n","                if fields[2] == \"gene\":\n","                    gene_id = fields[8].split(\";\")[0].split(\"=\")[1]\n","                    start = int(fields[3])\n","                    end = int(fields[4])\n","                    strand = fields[6]\n","                    if strand == \"+\":\n","                        promoter_start = max(1, start - upstream_length)\n","                        promoter_end = start - 1\n","                    elif strand == \"-\":\n","                        promoter_start = end + 1\n","                        promoter_end = min(len(genome_dict[fields[0]]), end + upstream_length)\n","                    gene_dict[gene_id] = {\"start\": promoter_start, \"end\": promoter_end, \"strand\": strand}\n","    \n","    # Extract promoter sequences\n","    with open(output_file, \"w\") as f:\n","        for gene_id in gene_dict:\n","            start = gene_dict[gene_id][\"start\"]\n","            end = gene_dict[gene_id][\"end\"]\n","            strand = gene_dict[gene_id][\"strand\"]\n","            if strand == \"+\":\n","                promoter_seq = genome_dict[fields[0]][start-1:end]\n","            elif strand == \"-\":\n","                promoter_seq = reverse_complement(genome_dict[fields[0]][start-1:end])\n","            f.write(f\">{gene_id}\\n{promoter_seq}\\n\")\n"],"metadata":{"id":"jD3A9-duHJXu"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"odJZMyGhzG5p"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def reverse_complement(seq):\n","    complement = {\"A\": \"T\", \"C\": \"G\", \"G\": \"C\", \"T\": \"A\", \"N\": \"N\"}\n","    seq = seq.upper()\n","    return \"\".join(complement[base] for base in reversed(seq))\n"],"metadata":{"id":"t1H4qvErIFNV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["extract_promoter_sequences(\"/content/gene_annotation.gff3\", \"/content/genome_assembly.fa\", 2000, \"output_file.fasta\")\n"],"metadata":{"id":"xPUFp06UHTeW"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["import re"],"metadata":{"id":"NxwvWfd1QR_7"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def find_dreb1a_binding_sites(promoter_file, motif_file, output_file):\n","    # Read promoter sequences from input file\n","    with open(promoter_file, \"r\") as f:\n","        promoter_seqs = f.readlines()\n","\n","    # Search for DREB1A binding site in each promoter sequence\n","    with open(output_file, \"w\") as f:\n","        for i in range(0, len(promoter_seqs), 2):\n","            promoter_id = promoter_seqs[i].strip()[1:]\n","            promoter_seq = promoter_seqs[i+1].strip()\n","            binding_sites = [match.start() for match in re.finditer(r\"CACCGAC\", promoter_seq)]\n","\n","            # Write binding sites to output file\n","            if binding_sites:\n","                f.write(f\">{promoter_id}\\n\")\n","                for site in binding_sites:\n","                    f.write(f\"{site+1}\\t{promoter_seq[site:site+7]}\\n\")\n"],"metadata":{"id":"s4abDrI4QTfV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["find_dreb1a_binding_sites(\"/content/output_file.fasta\", \"/content/MA0971.1.meme\", \"downstream.fasta\")\n"],"metadata":{"id":"1tDF6oWcQ5Is"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def identify_dreb1a_binding_sites(promoter_file, motif_file, output_file):\n","    # Read promoter sequences from input file\n","    with open(promoter_file, \"r\") as f:\n","        promoter_seqs = f.readlines()\n","    with open(motif_file, \"r\") as f:\n","        motif_seq = f.readlines()[1].strip()\n","# Search for motif in each promoter sequence\n","    with open(output_file, \"w\") as f:\n","        for i in range(0, len(promoter_seqs), 2):\n","            promoter_id = promoter_seqs[i].strip()[1:]\n","            promoter_seq = promoter_seqs[i+1].strip()\n","            motif_sites = [match.start() for match in re.finditer(motif_file, promoter_seq)]\n","\n","            # Write motif sites to output file\n","            if motif_sites:\n","                f.write(f\">{promoter_id}\\n\")\n","                for site in motif_sites:\n","                    f.write(f\"{site+1}\\t{promoter_seq[site:site+len(motif_seq)]}\\n\")"],"metadata":{"id":"L6-egREYTngs"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["identify_dreb1a_binding_sites(\"/content/output_file.fasta\", \"/content/MA0971.1.meme\", \"downstream2.fasta\")"],"metadata":{"id":"DYxnOG60UDB1"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":[],"metadata":{"id":"CSb7bp3TQRz6"}},{"cell_type":"code","source":["from Bio import motifs, SeqIO\n","from Bio.Alphabet import generic_dna\n","\n","def search_pwm_in_sequence(seq_file, pwm_file):\n","    # Load the DREB1A PWM from a MEME file\n","    with open(pwm_file) as f:\n","        m = motifs.parse(f, \"MEME\")\n","        pwm = m[0].pwm\n","\n","    # Load the promoter sequence\n","    promoter = SeqIO.read(seq_file, \"fasta\").seq\n","\n","    # Create a Bio.Seq object with a DNA alphabet\n","    seq = promoter.upper()\n","    seq = seq.transcribe()  # convert T to U for RNA sequences\n","    seq = seq.back_transcribe()  # convert U back to T for DNA sequences\n","    seq = seq.reverse_complement()  # reverse complement the sequence if necessary\n","    seq = seq.upper()\n","    seq = seq.tomutable()\n","    seq.alphabet = generic_dna\n","\n","    # Search for the motif in the promoter sequence\n","    hits = pwm.search(seq)\n","\n","    # Get the sequences and positions of the hits\n","    hit_seqs = []\n","    for hit in hits:\n","        start = hit[0]\n","        end = hit[1]\n","        score = hit[2]\n","        hit_seq = seq[start:end]\n","        hit_seq_str = str(hit_seq)\n","        hit_seqs.append((start, end, hit_seq_str))\n","\n","    return hit_seqs\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":564},"id":"iQuBvRhbu2DB","executionInfo":{"status":"error","timestamp":1678975459709,"user_tz":-60,"elapsed":1337,"user":{"displayName":"lateefat arogundade","userId":"11826085879239467444"}},"outputId":"bdd1fed6-999d-4a7b-8237-27f0fb0a591d"},"execution_count":null,"outputs":[{"output_type":"error","ename":"ImportError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)","\u001b[0;32m<ipython-input-3-795e06b46ac7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mmotifs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAlphabet\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mgeneric_dna\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0msearch_pwm_in_sequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpwm_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;31m# Load the DREB1A PWM from a MEME file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/Bio/Alphabet/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \"\"\"\n\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m raise ImportError(\n\u001b[0m\u001b[1;32m     21\u001b[0m     \u001b[0;34m\"Bio.Alphabet has been removed from Biopython. In many cases, the alphabet can simply be ignored and removed from scripts. In a few cases, you may need to specify the ``molecule_type`` as an annotation on a SeqRecord for your script to work correctly. Please see https://biopython.org/wiki/Alphabet for more information.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m )\n","\u001b[0;31mImportError\u001b[0m: Bio.Alphabet has been removed from Biopython. In many cases, the alphabet can simply be ignored and removed from scripts. In a few cases, you may need to specify the ``molecule_type`` as an annotation on a SeqRecord for your script to work correctly. Please see https://biopython.org/wiki/Alphabet for more information.","","\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"],"errorDetails":{"actions":[{"action":"open_url","actionText":"Open Examples","url":"/notebooks/snippets/importing_libraries.ipynb"}]}}]}]}